lms/views/basic_lti_launch.py

```diff
@@ -335,27 +310,64 @@ def _set_via_url(self, document_url):
         self.context.js_config["urls"].update(
             {"via_url": via_url(self.request, document_url)}
         )
-        self._set_submission_param("document_url", document_url)
+        self.set_canvas_submission_param("document_url", document_url)
+
+    # ---------------------------------------------------------------------- #
+    # Canvas specific functions
 
-    def _set_submission_param(self, name, value):
+    def is_launched_by_canvas(self):
+        return (
+            self.request.params.get("tool_consumer_info_product_family_code")
+            == "canvas"
+        )
+
+    def initialise_canvas_submission_params(self):
+        """
+        Add config used by frontend to call Canvas `record_submission` API.
+
+        The outcome reporting params are typically only available when
+        students (not teachers) launch an assignment.
```

Comment:
This seems like a code comment rather than a docstring--it doesn't seem to be describing what the method does but rather a detail of how it works. Making it a code comment would also allow it to be placed near the `lis_result_sourcedid` and `lis_outcome_service_url` params that it refers to, currently it might not be clear what "outcome reporting params" refers to. It'd be a bit more verbose but I think more clear if the code were written like this:

```python
    def initialise_canvas_submission_params(self):
        """Add config used by frontend to call Canvas `record_submission` API."""

        lis_result_sourcedid = self.request.params.get("lis_result_sourcedid")
        lis_outcome_service_url = self.request.params.get("lis_outcome_service_url")

        # Don't try to submit assignments to Canvas when the assignment is
        # launched by a teacher, or when the submission-related params are
        # missing for some other reason.
        if (not lis_result_sourcedid) or (not lis_outcome_service_url):
            return

        self.context.js_config["submissionParams"] = {
            "h_username": self.context.h_user.username,
            "lis_result_sourcedid": lis_result_sourcedid,
            "lis_outcome_service_url": lis_outcome_service_url,
        }
```

Reply:
It was taken from a comment before. I've used your comment, but I don't think the early return is required for a single statement.